# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.

import os
import pathlib
import pandas as pd
from snakemake.remote.S3 import RemoteProvider as S3RemoteProvider
from snakemake.utils import validate

S3 = S3RemoteProvider()

sys.path.insert(0, ".")
from lib import target_construction as tc


configfile: "config/config.yaml"


validate(config, "../schema/global_config_schema.yaml")


manifest = config["manifest"]
manifest = pd.read_csv(manifest, sep="\t")
reference_build = config["genome-build"]
octopus_num_intervals = 20

TARGETS = [
    expand(
        "results/multiqc/{projectid}/multiqc.{runtype}.html",
        projectid=list(manifest["projectid"].unique()),
        runtype=["fastq", "alignment"],
    )
]


rule all:
    input:
        TARGETS,


include: "rules/acquire_data.smk"
include: "rules/alignstats.smk"
include: "rules/bwamem2.smk"
include: "rules/duphold.smk"
include: "rules/fastp.smk"
include: "rules/fastqc.smk"
include: "rules/manta.smk"
include: "rules/mosdepth.smk"
include: "rules/multiqc.smk"
include: "rules/picard.smk"
include: "rules/references.smk"
include: "rules/somalier.smk"

include: "rules/verifybamid2.smk"
